packages/tpetra/core/src/Tpetra_MultiVector_def.hpp

```diff
@@ -3165,69 +3201,35 @@ namespace Tpetra {
       "this->getNumVectors() = " << numVecs << " != A.getNumVectors() = "
       << A.getNumVectors () << ".");
 
-    const impl_scalar_type theAlpha = static_cast<impl_scalar_type> (alpha);
-    const impl_scalar_type theBeta = static_cast<impl_scalar_type> (beta);
-    const std::pair<size_t, size_t> rowRng (0, lclNumRows);
-    const std::pair<size_t, size_t> colRng (0, numVecs);
-
     typedef typename dual_view_type::t_dev dev_view_type;
     typedef typename dual_view_type::t_host host_view_type;
 
-    // If we need sync to device, then host has the most recent version.
-    const bool useHostVersion = this->template need_sync<device_type> ();
-    if (useHostVersion) {
-      // Work on host, where A's data were most recently modified.  A
-      // is a "guest" of this method, so it's more polite to sync
-      // *this, than to sync A.
-      typedef typename host_view_type::memory_space cur_memory_space;
+    // Are this and A in the "host" state?
+    const bool thisIsHost = this->template need_sync<device_type> ();
+    const bool AIsHost = A.template need_sync<device_type> ();
 
+    // If they're not in the same state... sync *this because A is const
+    if(thisIsHost != AIsHost && AIsHost) {
```

Comment:
I have to think carefully about order of operations here.... I think we can simplify to two cases, plus minor modifications.  How about this:
```
using dev_mem_space = typename device_type::memory_space;
// with CudaUVMSpace, the host mirror has the same memory space;
// we need to distinguish between the two in order to force sync etc.
using host_mem_space = Kokkos::HostSpace;

if (! AIsHost) { // A is device, so run on device
  if (thisIsHost) {
    this->template sync<dev_mem_space> ();
  }
  this->template modify<dev_mem_space> ();
  this->template updateImpl<dev_mem_space> (alpha, A, beta);
}
else { // A is host, so run on host
  if (! thisIsHost) {
    this->template sync<host_mem_space> ();
  }
  this->template modify<host_mem_space> ();
  this->template updateImpl<host_mem_space> (alpha, A, beta);
}
```